### PR TITLE
feat: implement integration with dash0 cli for authentication

### PR DESCRIPTION
**Pull request by gat786**
Solves #65 

## Support using Dash0 CLI Configuration files for setting up provider authentication

This PR

* Introduced usage of two new environment variables

  `DASH0_API_URL` and `DASH0_CONFIG_DIR`

  `DASH0_API_URL` is introduced as a replacement for `DASH0_URL` we want to move away from the naming of `DASH0_URL` towards the newer one, we encourage using  `DASH0_API_URL` instead of `DASH0_URL`.

  `DASH0_CONFIG_DIR` is introduced to make it possible for you to define a 
custom directory from which the provider will try to load dash0 CLI config files from.
Introducing this variable also enabled us to write tests that do not operate on
users home directory.

* Adds support for authentication to dash0 using CLI configuration files.
Introduces a new optional provider attribute called `profile`.

  To build an API Client Provider will load credentials in the following precedence
  - Try to load `DASH0_API_URL` or DASH0_URL and DASH0_AUTH_TOKEN from environment variables.
  - If not found will try to load values from attributes configured i.e. `url`
  or `auth_token` attributes.
  - If still not found, will read the `profile` attribute and try to find a
  dash0 CLI profile with the same name and load the missing `url` or
  `auth_token` from it.
  - If an empty provider block was defined the provider will check if a dash0
  CLI configuration directory exists and an `activeProfile` is defined if it is
  it will load the missing values of `url` or `auth_token` from it.

### CODE REVIEW

**Comment by gat786** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by mmanciop** [comment]
@gat786 are you planning to write tests and a changelog for this feature?

File: internal/provider/provider.go [hunk 1]
```diff
@@ -86,6 +98,38 @@ func (p *dash0Provider) Configure(ctx context.Context, req provider.ConfigureReq
 		authToken = cfg.AuthToken.ValueString()
 	}
 
+	profile := ""
```
Comment (mmanciop) [inline internal/provider/provider.go]: If the profile is not found, the configuration is invalid, and the provider should fail

**Review by mmanciop — CHANGES_REQUESTED** [review]

File: internal/provider/provider.go (hunk 1 above)
Comment (gat786) [reply internal/provider/provider.go]: Even when url and auth_token are present?

**Comment by gat786** [comment]
Thats a good suggestion, I should start by writing Tests, it will make it easier for me to gauge whether the code I have written is correct or not. I will do that @mmanciop

**Comment by gat786** [comment]
I can write a changelog as well, I will have to see how it is done in this repo, I do not know as of yet.

**Comment by gat786** [comment]
@mmanciop I did implement the provider loading config from dash0 profiles according to what i think was correct, I also added some initial tests which verify the loading of missing values from profiles. 

I would appreciate a quick view and your thoughts, thanks!

File: internal/provider/provider.go [hunk 2]
```diff
@@ -86,20 +185,70 @@ func (p *dash0Provider) Configure(ctx context.Context, req provider.ConfigureReq
...
 			"Missing Dash0 URL",
 			"The provider cannot create the Dash0 API client because no Dash0 URL was provided. "+
-				"Set the `url` attribute in the provider block or set the DASH0_URL environment variable.",
+				"Set the `url` attribute in the provider block or set the DASH0_URL environment variable."+
+				" If not found in env variables or provider attributes, the provider will try to load value"+
+				" using `profile` defined in the provider attributes. Provider will load the missing value"+
+				" from configured dash0 CLI profile having the same name."+
+				" If `profile` is also not defined then the provider will try to load it using dash0 CLI's"+
+				" activeProfile defined at ~/.dash0/activeProfile.\n"+
+				exceptionHandled.Error(),
```
Comment (mmanciop) [inline internal/provider/provider.go]: If url/authToken are empty but the CLI config path succeeds, exceptionHandled is never assigned — yet `.Error()` is called on it unconditionally in the error messages. This will panic.

File: internal/provider/provider.go [hunk 3]
```diff
@@ -86,20 +185,70 @@ func (p *dash0Provider) Configure(ctx context.Context, req provider.ConfigureReq
...
 			"Missing Dash0 Auth Token",
 			"The provider cannot create the Dash0 API client because no Dash0 auth token was provided. "+
-				"Set the `auth_token` attribute in the provider block or set the DASH0_AUTH_TOKEN environment variable.",
+				"Set the `auth_token` attribute in the provider block or set the DASH0_AUTH_TOKEN environment variable.\n"+
+				" If not found in env variables or provider attributes, the provider will try to load value"+
+				" using `profile` defined in the provider attributes. Provider will load the missing value"+
+				" from configured dash0 CLI profile having the same name."+
+				" If `profile` is also not defined then the provider will try to load it using dash0 CLI's"+
+				" activeProfile defined at ~/.dash0/activeProfile.\n"+
+				exceptionHandled.Error(),
```
Comment (mmanciop) [inline internal/provider/provider.go]: If url/authToken are empty but the CLI config path succeeds, exceptionHandled is never assigned — yet `.Error()` is called on it unconditionally in the error messages. This will panic.

File: internal/provider/provider.go [hunk 4]
```diff
@@ -64,6 +92,77 @@ func (p *dash0Provider) Schema(_ context.Context, _ provider.SchemaRequest, resp
...
+	if dash0ActiveProfileFilePathExistsErr != nil {
+		// error stating activeProfileFilePath
+		return "", dash0ActiveProfileFilePathExistsErr
+
+	} else {
+		activeProfileFileContent,
+			activeProfileFileContentErr := os.ReadFile(dash0ActiveProfileFilePath)
+		if activeProfileFileContentErr != nil {
+			// error reading activeProfileFilePath
+			return "", dash0ActiveProfileFilePathExistsErr
```
Comment (mmanciop) [inline internal/provider/provider.go]: Should return `activeProfileFileContentErr`

File: internal/provider/provider.go [hunk 5]
```diff
@@ -86,20 +185,70 @@ func (p *dash0Provider) Configure(ctx context.Context, req provider.ConfigureReq
...
+		} else {
+			// Try to load values from dash0 CLI config files
+			if profile == "" {
+				var loadActiveProfileErr error
+				profile, loadActiveProfileErr = loadActiveProfileFile(homeDir)
+				if loadActiveProfileErr != nil {
+					exceptionHandled = loadActiveProfileErr
+				}
+			}
+
```
Comment (mmanciop) [inline internal/provider/provider.go:191]: The profile fallback only triggers when `url == "" || authToken == ""`, but values loaded from the profile unconditionally overwrite both. If a user provides `auth_token` but not `url`, the profile's authToken will silently replace the user-supplied one.

File: internal/provider/provider.go [hunk 6]
```diff
@@ -38,6 +40,22 @@ type dash0Provider struct {
...
+}
+
+type profileModel struct {
+	Name          string      `json:"name"`
+	Configuration configModel `json:"configuration"`
+}
+
+type profilesConfigFile struct {
+	Profiles []profileModel `json:"profiles"`
 }
```
Comment (mmanciop) [inline internal/provider/provider.go:48]: Custom JSON types that public types in API client, should not redeclare these

File: internal/provider/provider.go [hunk 7]
```diff
@@ -46,11 +64,21 @@ func (p *dash0Provider) Metadata(_ context.Context, _ provider.MetadataRequest,
...
-// Schema defines the provider-level schema for configuration data.
-func (p *dash0Provider) Schema(_ context.Context, _ provider.SchemaRequest, resp *provider.SchemaResponse) {
-	resp.Schema = schema.Schema{
+func providerSchema() schema.Schema {
+	return schema.Schema{
 		Description: `The Dash0 provider allows you to manage resources on the [Dash0](https://www.dash0.com) observability platform, including dashboards, check rules, recording rules, recording rule groups, synthetic checks, and views. Authentication can be provided via provider configuration attributes or via the DASH0_URL and DASH0_AUTH_TOKEN environment variables.`,
 		Attributes: map[string]schema.Attribute{
+			"profile": schema.StringAttribute{
+				Optional:  true,
+				Sensitive: true,
```
Comment (mmanciop) [inline internal/provider/provider.go]: A profile name is not a secret — marking it sensitive hides it in plan output for no reason.

File: internal/provider/provider_test.go [hunk 8]
```diff
@@ -12,6 +14,55 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// Only creates a temporary dash0 Config Directory for tests
+// You can safely run it on your local machine if the directory exists, it will
+// not overwrite
+func createTemporaryDash0CliConfig(t *testing.T) bool {
```
Comment (mmanciop) [inline internal/provider/provider_test.go]: This function writes to the actual home directory. This breaks the setup of anybody running tests locally. This makes tests non-hermetic and could destroy a developer's real CLI config on cleanup.

File: internal/provider/provider_test.go [hunk 9]
```diff
@@ -304,6 +311,215 @@ func TestDash0Provider_Configure_MissingBoth(t *testing.T) {
...
+	resp := &provider.ConfigureResponse{}
+
+	p.Configure(context.Background(), req, resp)
+
+	assert.False(t, resp.Diagnostics.HasError())
+	t.Log(resp.Diagnostics.Errors())
+	assert.NotNil(t, resp.ResourceData)
+	assert.NotNil(t, resp.DataSourceData)
+
+	if tempDirCreated {
```
Comment (mmanciop) [inline internal/provider/provider_test.go]: This pattern is fragile — if a test fails early, cleanup is skipped. Use `t.Cleanup()`.

File: internal/provider/provider.go [hunk 10]
```diff
@@ -86,20 +185,70 @@ func (p *dash0Provider) Configure(ctx context.Context, req provider.ConfigureReq
...
+		}
+	}
+
 	// Validate
 	if url == "" {
 		resp.Diagnostics.AddError(
 			"Missing Dash0 URL",
 			"The provider cannot create the Dash0 API client because no Dash0 URL was provided. "+
-				"Set the `url` attribute in the provider block or set the DASH0_URL environment variable.",
+				"Set the `url` attribute in the provider block or set the DASH0_URL environment variable."+
```
Comment (mmanciop) [inline internal/provider/provider.go]: The long multi-paragraph error strings duplicate the profile attribute documentation. Keep error messages concise and actionable.

File: internal/provider/provider.go [hunk 11]
```diff
@@ -86,20 +185,70 @@ func (p *dash0Provider) Configure(ctx context.Context, req provider.ConfigureReq
...
+				exceptionHandled.Error(),
 		)
 	}
 
 	if authToken == "" {
 		resp.Diagnostics.AddError(
 			"Missing Dash0 Auth Token",
 			"The provider cannot create the Dash0 API client because no Dash0 auth token was provided. "+
-				"Set the `auth_token` attribute in the provider block or set the DASH0_AUTH_TOKEN environment variable.",
+				"Set the `auth_token` attribute in the provider block or set the DASH0_AUTH_TOKEN environment variable.\n"+
```
Comment (mmanciop) [inline internal/provider/provider.go]: The long multi-paragraph error strings duplicate the profile attribute documentation. Keep error messages concise and actionable.

File: test/roundtrip/test_provider_empty_attributes.sh [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
```
Comment (mmanciop) [inline test/roundtrip/test_provider_empty_attributes.sh:1]: this test is incomplete and the file has misleading comments about keep_firing_for (copy-paste from another test) and only calls write_provider_tf with empty env vars. It doesn't assert anything.

File: internal/provider/provider.go [hunk 13]
```diff
@@ -64,6 +92,77 @@ func (p *dash0Provider) Schema(_ context.Context, _ provider.SchemaRequest, resp
...
+	dash0ActiveProfileFilePath := fmt.Sprintf(
+		"%s/.dash0/activeProfile",
+		homeDir,
+	)
+	_, dash0ActiveProfileFilePathExistsErr := os.Stat(dash0ActiveProfileFilePath)
+	if dash0ActiveProfileFilePathExistsErr != nil {
+		// error stating activeProfileFilePath
+		return "", dash0ActiveProfileFilePathExistsErr
+
+	} else {
```
Comment (mmanciop) [inline internal/provider/provider.go]: No need for else if the IF branch returns. Not idiomatic go.

File: internal/provider/provider.go [hunk 14]
```diff
@@ -64,6 +92,77 @@ func (p *dash0Provider) Schema(_ context.Context, _ provider.SchemaRequest, resp
...
+		dash0ProfilesFileContent,
+			dash0ProfilesFileContentReadErr := os.ReadFile(dash0ProfilesFilePath)
+		if dash0ProfilesFileContentReadErr != nil {
+			return configModel{}, dash0ProfilesFileExistsErr
+		}
+		var profilesConfigFile profilesConfigFile
+		profileJsonUnmarshalErr := json.Unmarshal(dash0ProfilesFileContent, &profilesConfigFile)
+		if profileJsonUnmarshalErr != nil {
+			return configModel{}, profileJsonUnmarshalErr
+		} else {
```
Comment (mmanciop) [inline internal/provider/provider.go]: No need for else if the IF branch returns. Not idiomatic go.

**Review by mmanciop — CHANGES_REQUESTED** [review]
Besides the comments inline:

- No changelog entry (.chloggen/*.yaml) as required by the project.                                                                                     
- The profile attribute is not documented in examples/ or docs/.                                                                                      
- No test for the case where ~/.dash0/ doesn't exist at all and no credentials are provided (should still produce clean error messages, but would currently panic.

**Comment by gat786** [comment]
Thanks Michele, this is a very detailed breakdown of things that I should fix, I will do it soon and ask again.

File: internal/provider/provider.go (hunk 5 above)
Comment (gat786) [reply internal/provider/provider.go:191]: I replaced the code to only replace the value if current value for `url` is empty as well as the incoming value from profiles file is non-empty.

File: internal/provider/provider.go (hunk 6 above)
Comment (gat786) [reply internal/provider/provider.go:48]: I found about apiClient very late in my implementation, I will use the client instead of new definitions.

File: internal/provider/provider.go (hunk 7 above)
Comment (gat786) [reply internal/provider/provider.go]: copy/paste mistake from my end.

File: internal/provider/provider_test.go (hunk 8 above)
Comment (gat786) [reply internal/provider/provider_test.go]: This is something I was thinking about how to make it hermetic, but I did not find a way, since the dash0 CLI files are always in the home directory our code also reads from that directory inorder to test the full flow without using the home directory we will have to make the directory path generation step somehow configurable so that when needed we can use a preconfigured test directory and in production it uses `$HOME` do you think that is the correct way? 

By the way, the code which actually reads values from config directories is already separated into a separate function and it accepts `homeDir` as a parameter so in theory we could create a temporary directory pass it as the `homeDir` and test that particular function if need be. but testing the entire flow was something that I didn't know how to do.

File: internal/provider/provider_test.go (hunk 9 above)
Comment (gat786) [reply internal/provider/provider_test.go]: I agree.

File: internal/provider/provider_test.go (hunk 8 above)
Comment (gat786) [reply internal/provider/provider_test.go]: I believe allowing to choose a custom config dir path on basis of an environment variable will allow our test cases to work. For e.g. along with `DASH0_AUTH_TOKEN` and `DASH0_API_URL` we also read `DASH0_CONFIG_DIR` which if provided the provider will read config files from it. which if not provided it will read from the default home dir, the cli will continue to work as before and provider will also work but in our test cases we can create a tempDir using `t.TempDir` create the dummyconfig files in it, use it for tests and get it rid of it when the tests are done.

File: internal/provider/provider.go (hunk 10 above)
Comment (gat786) [reply internal/provider/provider.go]: I will try to sum it up in a few lines as better as I can but if you have an official template that we should use, send it my way and I will use that.

File: test/roundtrip/test_provider_empty_attributes.sh (hunk 12 above)
Comment (gat786) [reply test/roundtrip/test_provider_empty_attributes.sh:1]: I did not write these tests as of now, only the tests written in `provider_test.go` are implemented I thought I should get an idea of what you think of my implementation before I go and create those as well.

File: internal/provider/provider.go (hunk 13 above)
Comment (gat786) [reply internal/provider/provider.go]: I will fix.

File: internal/provider/provider.go (hunk 3 above)
Comment (gat786) [reply internal/provider/provider.go]: I now expand the exception handled before the actual error printing and in a conditional way, I only attach the value if it is not-nil

File: docs/index.md [hunk 15]
```diff
@@ -314,4 +331,4 @@ YAML
 # resource "dash0_notification_channel" "from_file" {
 #   notification_channel_yaml = file("${path.module}/notification_channel.yaml")
 # }
-```
\ No newline at end of file
+```
```
Comment (mmanciop) [inline docs/index.md]: Spurious diff, revert please

File: Makefile [hunk 16]
```diff
@@ -81,4 +81,4 @@ chlog-preview: $(CHLOGGEN)
 	$(CHLOGGEN) update --config .chloggen/config.yaml --dry
 
 chlog-update: $(CHLOGGEN)
-	$(CHLOGGEN) update --config .chloggen/config.yaml --version $(VERSION)
\ No newline at end of file
+	$(CHLOGGEN) update --config .chloggen/config.yaml --version $(VERSION)
```
Comment (mmanciop) [inline Makefile]: Spurious diff, revert please

File: internal/provider/provider.go (hunk 1 above)
Comment (mmanciop) [reply internal/provider/provider.go]: Yes ofc

File: .chloggen/provider-add-support-for-authn-using-cli-profiles.yaml [hunk 17]
```diff
@@ -20,9 +20,13 @@ subtext: |
   Adds support for authentication to dash0 using CLI configuration files.
   Introduces a new optional provider attribute called `profile`.
 
+  Introduces support for a new ENV variable `DASH0_CONFIG_DIR` which if specified
```
Comment (mmanciop) [inline .chloggen/provider-add-support-for-authn-using-cli-profiles.yaml]: What do you mean "New"? We have had it for ages :-)

File: .chloggen/provider-add-support-for-authn-using-cli-profiles.yaml [hunk 18]
```diff
@@ -31,6 +35,9 @@ subtext: |
   - If an empty provider block was defined the provider will check if a dash0
     CLI configuration directory exists and an `activeProfile` is defined if it is
     it will load the missing values of `url` or `auth_token` from it.
+  - The DASH0_CONFIG_DIR env variable if specified is supposed to be directory in which
```
Comment (mmanciop) [inline .chloggen/provider-add-support-for-authn-using-cli-profiles.yaml]: Remove this, the feature already exists and it already documented

**Review by mmanciop — APPROVED** [review]
